doc/spec/json.md

```diff
@@ -0,0 +1,87 @@
+# Docker Distribution JSON Canonicalization
+
+## Introduction
+
+To provide consistent content hashing of JSON object throughout Docker
+Distribution APIs, a canonical JSON format has been defined. The
+canonicalization is balance between conistency and readability. The format
+also provides consistent output for effective caching.
+
+## Rules
+
+1. Object keys shall always appear in lexically sorted order.
+2. The colon (":") shall always be followed by a single space.
+3. An indent of three spaces ("   ") shall be used.
+4. The number of indents used shall be proportional to the current depth of
+   nesting, where the depth increments after the start of an array or object
+   and decrements afterwards.
+5. The comma (",") shall always be followed by a newline and the number of
+   indents corresponding to the current depth.
+6. Empty objects ("{}") and arrays ("[]") shall not have any internal spacing.
+7. No "trailing commas" are allowed in object or array definitions.
+
+## Examples
+
+The following is a simple example of a canonicalized JSON string:
+
+```json
+{
+   "asdf": 1,
+   "qwer": [],
+   "zxcv": [
+      {},
+      true,
+      1000000000,
+      "tyui"
+   ]
+}
+```
+
+## Reference
+
+### Other Canonicalizations
+
+The OLPC project specifies [Canonical
+JSON](http://wiki.laptop.org/go/Canonical_JSON). While this is used in TUF,
+which may be used with other distribution-related protocols, this alternative
+format has been proposed for readability. Specifications complying with either
+this specification or an alternative should explicitly call out the
+canonicalization format.
+
+### Go
+
+In Go, the canonical JSON format spacing can be obtained by calling the
+[`json.Indent`](http://golang.org/src/encoding/json/indent.go?s=1918:1989#L65)
+with the following arguments:
+
+```go
+incoming := getBytes()
+var canonical bytes.Buffer
+if err := json.Indent(&canonical, incoming, "", "   "); err != nil {
+	// ... handle error
+}
+```
+
+By default, [`encoding/json`](http://golang.org/pkg/encoding/json/) will
+generate objects with sorted keys. Simply using `json.MarshalIndent` with the
```

Comment:
@dmcgowan This effectively just allows for another specification to call out a field order that makes sense for readability that is not lexically sorted and still be compliant with this specification. We already effectively have this for manifests and other types will follow (`Event` comes to mind).

When unmarshaling, the key order doesn't matter. This specification only applies to the generated format. When marshaling, the code needs to take appropriate care that the keys have a specific order, be it lexically sorted or based on a specification. The goal is to have the keys be marshaled in a consistent order.

Perhaps, we could clarify the wording. Suggestions are welcome.